src/Polysemy/Internal/Union.hs

```diff
@@ -56,77 +59,79 @@ import Polysemy.Internal.CustomErrors
-       , weaveState :: f ()
-         -- ^ A piece of state that other effects' interpreters have already
-         -- woven through this 'Weaving'. @f@ is a 'Functor', so you can always
-         -- 'fmap' into this thing.
-       , weaveDistrib :: forall x. f (m x) -> n (f x)
-         -- ^ Distribute @f@ by transforming @m@ into @n@. We have invariants
-         -- on @m@ and @n@, which means in actuality this function looks like
-         -- @f ('Polysemy.Sem' (Some ': Effects ': r) x) -> 'Polysemy.Sem' r (f
-         -- x)@.
-       , weaveResult :: f a -> b
-         -- ^ Even though @f a@ is the moral resulting type of 'Weaving', we
-         -- can't expose that fact; such a thing would prevent 'Polysemy.Sem'
-         -- from being a 'Monad'.
-       , weaveInspect :: forall x. f x -> Maybe x
-         -- ^ A function for attempting to see inside an @f@. This is no
-         -- guarantees that such a thing will succeed (for example,
-         -- 'Polysemy.Error.Error' might have 'Polysemy.Error.throw'n.)
-       }
-    -> Weaving e n b
+    :: forall f e rBefore a resultType mAfter. (Functor f)
+    => !(WeavingDetails f e rBefore a resultType mAfter)
+    -> Weaving e mAfter resultType
+
+data Functor f => WeavingDetails f e rBefore a resultType mAfter =
+  WeavingDetails {
+    weaveEffect :: e (Sem rBefore) a
+    -- ^ The original effect GADT originally lifted via
+    -- 'Polysemy.Internal.send'.
+  , weaveState :: f ()
+    -- ^ A piece of state that other effects' interpreters have already
+    -- woven through this 'Weaving'. @f@ is a 'Functor', so you can always
+    -- 'fmap' into this thing.
+  , weaveDistrib :: forall x. f (Sem rBefore x) -> mAfter (f x)
+    -- ^ Distribute @f@ by transforming @Sem rBefore@ into @mAfter@. We have an
+    -- invariants on @rBefore@, which means in actuality this function looks like
+    -- @f ('Polysemy.Sem' rBefore x) -> mAfter (f x)@.
+  , weaveResult :: f a -> resultType
+    -- ^ Even though @f a@ is the moral resulting type of 'Weaving', we
+    -- can't expose that fact; such a thing would prevent 'Polysemy.Sem'
+    -- from being a 'Monad'.
+  , weaveInspect :: forall x. f x -> Maybe x
+    -- ^ A function for attempting to see inside an @f@. This is no
+    -- guarantees that such a thing will succeed (for example,
+    -- 'Polysemy.Error.Error' might have 'Polysemy.Error.throw'n.)
+  }
 
 instance Functor (Weaving e m) where
-  fmap f (Weaving e s d f' v) = Weaving e s d (f . f') v
+  fmap f (Weaving (WeavingDetails e s d f' v)) =
+    Weaving $ WeavingDetails e s d (f . f') v
   {-# INLINE fmap #-}
 
 
 
 weave
-    :: (Functor s, Functor m, Functor n)
+    :: (Functor s, Functor mMid, Functor mAfter)
     => s ()
-    -> (∀ x. s (m x) -> n (s x))
+    -> (∀ x. s (mMid x) -> mAfter (s x))
     -> (∀ x. s x -> Maybe x)
-    -> Union r m a
-    -> Union r n (s a)
-weave s' d v' (Union w (Weaving e s nt f v)) = Union w $
-    Weaving e (Compose $ s <$ s')
```

Comment:
While I appreciate the sentiment of having more descriptive variable names, I don't think they're particularly helpful in this instance. They add clutter without providing much additional information.
In particular, I'm not sure whether or not `mMid` helps or hurts. It's true in the sense that it is the middle monad between the initial existential monad and the monad we're `weave`-ing to, but there's never any reason to care about the initial existential monad beyond the fact that it's a monad. Conceptually, `mMid` represents the monad associated to the original Union, rather than something intermediary.

My opinion is that you leave `weave` unchanged.

Reply:
Changed back to `m` and `n`.